src/player.c

```diff
@@ -496,3 +531,111 @@ void *BarPlayerThread (void *data) {
 	return (void *) pret;
 }
 
+void *BarAoPlayThread (void *data) {
+	assert (data != NULL);
+
+	player_t *const player = data;
+
+	AVFrame *frm = NULL;
+	AVPacket *pkt = NULL;
+
+	bool quit = false;
+
+	while (!shouldQuit (player) && !quit) {
+		struct _QUEUE_ELEMT * queue_element = NULL;
+		while (!shouldQuit (player) && !quit && queue_element == NULL) {
+
+			/* force pausing until the queue size get 256 or decoding is done*/
+			pthread_mutex_lock (&player->aoplay_lock);
+			if (player->forcePause && !player->queueDone) {
+				do {
+					pthread_cond_wait (&player->aoplay_cond, &player->aoplay_lock);
+				} while (player->forcePause && !player->queueDone);
+			}
+			pthread_mutex_unlock (&player->aoplay_lock);
+
+			/* pausing by user*/
+			pthread_mutex_lock (&player->lock);
+			if (player->doPause) {
+				av_read_pause (player->fctx);
+				do {
+					pthread_cond_wait (&player->cond, &player->lock);
+				} while (player->doPause);
+				av_read_play (player->fctx);
+			}
+			pthread_mutex_unlock (&player->lock);
+
+			pthread_mutex_lock (&player->aoplay_lock);
+			queue_element = STAILQ_FIRST (&player->queue_head);
+			pthread_mutex_unlock (&player->aoplay_lock);
+		}
+
+		if (queue_element == NULL) {
+			break;
+		}
+
+		int numChannels;
```

Comment:
Just a stylistic thing: You can declare variables (const please) inside switch cases if you’re enclosing the case’s body in brackets like this:

```c
switch (foo) {
     case bar: {
         const int baz = 123;
     }
}
```